builder/dockerfile/dispatchers.go

```diff
@@ -23,18 +23,66 @@ import (
 	"github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/api/types/strslice"
 	"github.com/docker/docker/builder"
-	"github.com/docker/docker/builder/dockerfile/parser"
+	"github.com/docker/docker/builder/dockerfile/command"
 	"github.com/docker/docker/pkg/signal"
 	"github.com/docker/go-connections/nat"
 	"github.com/pkg/errors"
 )
 
+func copyEnv(previous []string) []string {
+	result := make([]string, len(previous))
+	copy(result, previous)
+	return result
+}
+
+type baseCommand struct {
+	builder         *Builder
+	dispatchMessage string
+}
+
+func (c *baseCommand) writeDispatchMessage() {
+	fmt.Fprintln(c.builder.Stdout, c.dispatchMessage)
+}
+
+func dispatchEnv(state *dispatchState, c *command.EnvCommand, expand processWordFunc) error {
+	runConfig := state.runConfig
+	commitMessage := bytes.NewBufferString("ENV")
+	for _, e := range c.Env {
+		name, err := processWordSingleOutput(e.Key, expand)
```

Comment:
The idea here is to mutate the typed command structs when doing variable expansion ? Does it fit the idea of having a stateless ParseResult ? (but anyway, for execution parallelism, we will certainly need a Expand pass between the Parse and Dispatch passes so I suppose it is right)